### ISSUE TITLE
Trouble exporting roles in config due to magic ID prefixing

### ISSUE THREAD

**Issue by pfrenssen** [amitaibu/og#216]
Currently when a new `OgRole` is saved, the role ID gets prefixed with the group entity type ID and the group bundle ID in `OgRole::save()`:

```
  public function save() {
    if ($this->isNew()) {
      // When assigning a role to group we need to add a prefix to the ID in
      // order to prevent duplicate IDs.
      $prefix = $this->getGroupType() . '-' . $this->getGroupBundle() . '-';

      if (!empty($this->getGroupId())) {
        $prefix .= $this->getGroupId() . '-';
      }

      $this->id = $prefix . $this->id();
    }

    parent::save();
  }
```

This however causes problems in install profiles and modules supplying roles in config. Whenever a profile or module is installed, a new OgRole is saved with the ID that is read from the YAML file which will already contain the prefix, causing it to be prefixed twice. This can be worked around by manually editing the YAML file, but then the problem will reappear on a subsequent config exports.

This will probably also happen when moving config from staging to production, since the same cycle will occur.

We probably should not do any magic prefixing. It would be better to require the correct ID to be set and check this in OgRole::save().

**Comment by pfrenssen** [comment]
Maybe instead of calling `setId()` with the role name we can provide a `setName()` method to set it.  Then when saving we can generate the "fully qualified" ID if none has been set. Config can then still use the standard `setId()` method to set the FQID.

**Comment by amitaibu** [comment]
The advantage of the prefixing is that it also ensures unique roles. Not sure I get your `setName` idea.

**Comment by pfrenssen** [comment]
The thing we are now setting as the ID is not the ID (which needs to be unique) but it is the role name. So we should not use `setId()` for setting the role name, we should use `setName()` for it. Then when we save the entity:
- If we do not have an ID yet, we can generate a unique ID from concatenating the entity ID, the bundle ID and the role name, thus guaranteeing a unique ID as we have now.
- The fully qualified ID can still be set with `setId()`, but you will have to pass the unique ID. This will ensure that `ConfigInterface` still works (as well as all other code that thinks that `setId()` is in fact used to set the ID, instead of some intermediate format which will be altered on save).
- We can get the role name back by providing a `getName()` method, which is now impossible.

**Comment by amitaibu** [comment]
👍

**Comment by pfrenssen** [comment]
A proposed fix is included in #237.
